File: services/alp-dataflow-gen/pysrc/api/OpenIdAPI.py
import requests
import os
import jwt
from datetime import datetime


class OpenIdAPI:
    def __init__(self):
        if os.getenv('IDP__ISSUER_URL') is None:
            raise ValueError("IDP__ISSUER_URL is undefined")

        if os.getenv('IDP__ALP_DATA__CLIENT_ID') is None:
            raise ValueError("IDP__ALP_DATA__CLIENT_ID is undefined")

        if os.getenv('IDP__ALP_DATA__CLIENT_SECRET') is None:
            raise ValueError("IDP__ALP_DATA__CLIENT_SECRET is undefined")

        if os.getenv("PYTHON_VERIFY_SSL") == 'true' and os.getenv('TLS__INTERNAL__CA_CRT') is None:
            raise ValueError("TLS__INTERNAL__CA_CRT is undefined")

        self.url = os.getenv('IDP__ISSUER_URL')
        self.clientId = os.getenv('IDP__ALP_DATA__CLIENT_ID')
        self.clientSecret = os.getenv('IDP__ALP_DATA__CLIENT_SECRET')
        self.verifySsl = False if os.getenv(
            "PYTHON_VERIFY_SSL") == 'false' else os.getenv('TLS__INTERNAL__CA_CRT')

    def getOptions(self):
        return {
            "Content-Type": "application/json",
        }

    def getClientCredentialToken(self) -> str:
        params = {
            'grant_type': "client_credentials",
            'client_id': self.clientId,
            'client_secret': self.clientSecret,
        }

        result = requests.post(
            f"{self.url}/token",
            headers=self.getOptions(),
            verify=self.verifySsl,
            json=params
        )

        if ((result.status_code >= 400) and (result.status_code < 600)):
            raise Exception(
                f"OpenIdAPI Failed to get client credential token, {result.content}")
        else:
            return result.json()['access_token']

    def isTokenExpiredOrEmpty(self, token: str | None):
        if (not token):
            return True

        decodedToken = jwt.decode(token, options={'verify_signature': False})
        return int(datetime.now().timestamp()) > decodedToken['exp']

File: functions/alp-usermgmt/src/api/OpenIDAPI.ts
import { AxiosResponse } from 'axios'
import jwt from 'jsonwebtoken'
import https from 'https'
import { post } from './request-util'

interface IClientMetadata {
  issuerUrl: string
}

export interface IClientCredentials {
  clientId: string
  clientSecret: string
  scope: string
  resource?: string
}

interface ITokenResponse {
  access_token: string
}

export class OpenIDAPI {
  private readonly issuerUrl: string
  private readonly httpsAgent: https.Agent

  constructor({ issuerUrl }: IClientMetadata) {
    this.issuerUrl = issuerUrl.endsWith('/') ? issuerUrl : `${issuerUrl}/`

    this.httpsAgent = new https.Agent({
      rejectUnauthorized: this.issuerUrl.startsWith('https://alp-logto-') ? false : true
    })
  }

  async getClientCredentialsToken({ clientId, clientSecret, scope, resource }: IClientCredentials) {
    const params: any = {
      grant_type: 'client_credentials',
      client_id: clientId,
      client_secret: clientSecret,
      scope,
      resource
    }

    const body = Object.keys(params)
      .map(key => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`)
      .join('&')

    let result: AxiosResponse<ITokenResponse> | undefined
    try {
      result = await post<ITokenResponse>(`${this.issuerUrl}token`, body, {
        httpsAgent: this.httpsAgent,
        headers: {
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      })
    } catch (err) {
      console.error('Error when getting client credentials token', err)
    }

    return result?.data
  }

  isTokenExpiredOrEmpty(token?: string) {
    if (!token) {
      return true
    } else {
      const decodedToken = jwt.decode(token) as jwt.JwtPayload
      return decodedToken?.exp && decodedToken.exp < Date.now() / 1000
    }
  }
}
